src/core/crypto/publicKey.ts

```diff
@@ -0,0 +1,108 @@
+import { Serializable } from "../../bcs";
+import { AuthenticationKeyScheme, HexInput, SigningSchemeInput } from "../../types";
+import { AuthenticationKey } from "../authenticationKey";
+import { Hex } from "../hex";
+import { AccountSignature, CryptoSignature } from "./signature";
+
+/**
+ * Arguments for verifying a crypto signature
+ */
+export interface VerifyCryptoSignatureArgs {
+  message: HexInput;
+  signature: CryptoSignature;
+}
+
+/**
+ * Arguments for verifying an account signature
+ */
+export interface VerifyAccountSignatureArgs {
+  message: HexInput;
+  signature: AccountSignature;
+}
+
+/**
+ * An abstract representation of a crypto public key,
+ * associated to a specific signature scheme e.g. Ed25519 or Secp256k1
+ *
+ * Provides a common interface for verifying any crypto signature i.e. a signature
+ * associated to a specific signature scheme.
+ */
+export abstract class CryptoPublicKey extends Serializable {
+  /**
+   * The signature scheme associated with the public key.
+   *
+   * This property can be used at runtime to match
+   * the signature scheme of a given PrivateKey or CryptoSignature.
+   */
+  abstract readonly signatureScheme: SigningSchemeInput;
+
+  /**
+   * Verifies that the private key associated with this public key signed the message with the given signature.
+   * @param args.message The message that was signed
+   * @param args.signature The signature to verify
+   */
+  abstract verifySignature(args: VerifyCryptoSignatureArgs): boolean;
+
+  /**
+   * Get the raw public key bytes
+   */
+  abstract toUint8Array(): Uint8Array;
+
+  /**
+   * Get the public key as a hex string with a 0x prefix e.g. 0x123456...
+   */
+  toString(): string {
+    const bytes = this.toUint8Array();
+    return Hex.fromHexInput(bytes).toString();
+  }
+}
+
+/**
+ * An abstract representation of an account public key,
+ * associated to a specific authentication scheme e.g. Ed25519 or SingleKey
+ *
+ * Provides a common interface for verifying any `AccountSignature` i.e. a signature
+ * generated by an account and associated to a specific authentication scheme.
+ */
+export abstract class AccountPublicKey extends Serializable {
+  /**
+   * The authentication scheme associated with the public key.
+   *
+   * This property can be used at runtime to match
+   * the authentication scheme of a given AccountSignature.
+   */
+  abstract readonly authenticationScheme: AuthenticationKeyScheme;
```

Comment:
As discussed offline `AccountAuthenticator` wouldn't work imo, since you'd expect an `AccountAuthenticator` instance.

I agree all the schemes out there are really confusing and we should clean that up at some point.
Specifically SigninScheme vs SigningSchemeInput are conceptually very different but sound very similar.

`AuthenticationScheme` makes sense to me, as it's the scheme used to derive an authentication key from the public key bytes.

```
- AuthenticationScheme.Ed25519
Ed25519Account
Ed25519PublicKey
Ed25519Signature
AccountAuthenticatorEd25519

- AuthenticationScheme.MulitEd25519
MulitEd25519PublicKey
MulitEd25519Signature
AccountAuthenticatorMulitEd25519

- AuthenticationScheme.SingleKey
SingleKeyAccount
AnyPublicKey
AnySignature
AccountAuthenticatorSingleKey

- AuthenticationScheme.MultiKey
MultiKey
MultiKeySignature
AccountAuthenticatorMultiKey
```

But tying this concept to authenticators also would work.
We could use `AccountAuthenticatorVariant` which is the enum used for `AccountAuthenticator`.

Reply:
Something with `AccountAuthenticator` so it's consistent with the Rust codebase would be good. Will leave it up to you on what specifically you want to make it. For context, I had no idea what an `authenticationScheme` meant. I thought it was a `SignatureScheme` or like an HTTP authentication scheme which didn't make much sense to me. If you google "Authentication Scheme" you'll see what I mean

Reply:
+1 to `AccountAuthenticatorType`

Reply:
after some discussion, we opted to just drop the field